app/aboot/lk2nd-motorola.c

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <debug.h>
+#include <libfdt.h>
+#include <malloc.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include "lk2nd-device.h"
+#include "lk2nd-motorola.h"
+#include "smem.h"
+
+#define SMEM_ID_VENDOR0          0x86
+#define SMEM_KERNEL_RESERVE      SMEM_ID_VENDOR0
+#define SMEM_KERNEL_RESERVE_SIZE 1024
+
+void lk2nd_motorola_smem_write_unit_info(const void *fdt, int offset)
+{
+	int len = 0;
+	fdt_getprop(fdt, offset, "lk2nd,motorola-unitinfo", &len);
+	if (len == 0) {
+		return;
+	}
+
+	struct mmi_unit_info *unit_info;
+	unit_info = malloc(SMEM_KERNEL_RESERVE_SIZE);
+	memset(unit_info, 0, SMEM_KERNEL_RESERVE_SIZE);
+
+	int chosen_offset = fdt_path_offset(fdt, "/chosen");
+	const uint32_t *system_rev = fdt_getprop(fdt, chosen_offset, "linux,hwrev", NULL);
+	const uint32_t *system_serial_low = fdt_getprop(fdt, chosen_offset, "linux,seriallow", NULL);
+	const uint32_t *system_serial_high = fdt_getprop(fdt, chosen_offset, "linux,serialhigh", NULL);
+	const uint32_t *powerup_reason = fdt_getprop(fdt, chosen_offset, "mmi,powerup_reason", NULL);
+
+	const char *baseband = fdt_getprop(fdt, chosen_offset, "mmi,baseband", &len);
+	if (len == -1) {
+		baseband = "";
+	}
+
+	unit_info->version = MMI_UNIT_INFO_VER;
+	unit_info->system_rev = fdt32_to_cpu(*system_rev);
+	unit_info->system_serial_low = fdt32_to_cpu(*system_serial_low);
+	unit_info->system_serial_high = fdt32_to_cpu(*system_serial_high);
+	strlcpy(unit_info->machine, "", MACHINE_MAX_LEN);
+	strlcpy(unit_info->barcode, lk2nd_dev.serialno, BARCODE_MAX_LEN);
+	strlcpy(unit_info->baseband, baseband, BASEBAND_MAX_LEN);
+	strlcpy(unit_info->carrier, lk2nd_dev.carrier, CARRIER_MAX_LEN);
+	strlcpy(unit_info->device, lk2nd_dev.device, DEVICE_MAX_LEN);
+	unit_info->powerup_reason = fdt32_to_cpu(*powerup_reason);
+
+	const uint32_t radio = atoui(lk2nd_dev.radio);
+	unit_info->radio = radio;
+	if (radio == 0) {
+		strlcpy(unit_info->radio_str, lk2nd_dev.radio, RADIO_MAX_LEN);
+	}
+
+	dprintf(INFO, "Setting motorola unit info: version = 0x%02x, device = '%s', radio = 0x%x, radio_str = '%s', system_rev = 0x%04x, system_serial = 0x%08x%08x, machine = '%s', barcode = '%s', baseband = '%s', carrier = '%s', pu_reason = 0x%08x\n",
+		unit_info->version,
+		unit_info->device,
+		unit_info->radio,
+		unit_info->radio_str,
+		unit_info->system_rev,
+		unit_info->system_serial_high, unit_info->system_serial_low,
+		unit_info->machine, unit_info->barcode,
+		unit_info->baseband, unit_info->carrier,
+		unit_info->powerup_reason);
+
+	int ret = 0;
```

Comment:
lk kind of follows ancient C code standards where all variables are declared on top of the method. (like in the kernel)